File: csharp/969.pancake-sorting.cs
/*
 * @lc app=leetcode id=969 lang=csharp
 *
 * [969] Pancake Sorting
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static void Sort(int[] A, int end) {
        for (int i = 0; i < end; ++i, --end) {
            int tmp = A[i];
            A[i] = A[end];
            A[end] = tmp;
        }
    }

    public IList<int> PancakeSort(int[] A) {
        int n = A.Length;
        var ret = new List<int>();
        for (int i = 0; i < n; ++i) {
            int r = n - 1 - i, mi = 0, m = A[0];
            for (int j = 1; j <= r; ++j) {
                if (A[j] > m) {
                    mi = j;
                    m = A[j];
                }
            }
            Sort(A, mi);
            Sort(A, r);
            ret.Add(mi + 1);
            ret.Add(r + 1);
        }
        return ret;
    }
}
// @lc code=end

File: java/969.pancake-sorting.java
/*
 * @lc app=leetcode id=969 lang=java
 *
 * [969] Pancake Sorting
 */

import java.util.ArrayList;
import java.util.List;

// @lc code=start

class Solution {
    public List<Integer> pancakeSort(int[] arr) {
        int n = arr.length;
        List<Integer> ret = new ArrayList<>(n * 2);
        for (int i = 0; i < n; ++i) {
            int maxIndex = 0, endIndex = n - 1 - i;
            for (int j = 1; j <= endIndex; ++j) {
                if (arr[j] > arr[maxIndex]) {
                    maxIndex = j;
                }
            }
            if (maxIndex == endIndex) {
                continue;
            }
            if (maxIndex > 0) {
                flip(arr, maxIndex);
                ret.add(maxIndex + 1);
            }
            if (endIndex > 0) {
                flip(arr, endIndex);
                ret.add(endIndex + 1);
            }
        }
        return ret;
    }

    private static void flip(int[] arr, int endIndex) {
        for (int i = 0; 2 * i < endIndex; ++i) {
            int tmp = arr[i];
            arr[i] = arr[endIndex - i];
            arr[endIndex - i] = tmp;
        }
    }
}
// @lc code=end

